hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/CollectionTypes.java

```diff
@@ -0,0 +1,480 @@
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.concurrent.CompletionStage;
+
+import org.hibernate.Hibernate;
+import org.hibernate.HibernateException;
+import org.hibernate.collection.spi.AbstractPersistentCollection;
+import org.hibernate.collection.spi.PersistentArrayHolder;
+import org.hibernate.collection.spi.PersistentCollection;
+import org.hibernate.engine.spi.CollectionEntry;
+import org.hibernate.engine.spi.PersistenceContext;
+import org.hibernate.engine.spi.SessionImplementor;
+import org.hibernate.metamodel.mapping.PluralAttributeMapping;
+import org.hibernate.persister.collection.CollectionPersister;
+import org.hibernate.reactive.logging.impl.Log;
+import org.hibernate.reactive.logging.impl.LoggerFactory;
+import org.hibernate.reactive.util.impl.CompletionStages;
+import org.hibernate.type.ArrayType;
+import org.hibernate.type.CollectionType;
+import org.hibernate.type.CustomCollectionType;
+import org.hibernate.type.EntityType;
+import org.hibernate.type.ForeignKeyDirection;
+import org.hibernate.type.MapType;
+import org.hibernate.type.Type;
+
+import static org.hibernate.bytecode.enhance.spi.LazyPropertyInitializer.UNFETCHED_PROPERTY;
+import static org.hibernate.internal.util.collections.CollectionHelper.mapOfSize;
+import static org.hibernate.pretty.MessageHelper.collectionInfoString;
+import static org.hibernate.reactive.util.impl.CompletionStages.completedFuture;
+import static org.hibernate.reactive.util.impl.CompletionStages.loop;
+import static org.hibernate.reactive.util.impl.CompletionStages.voidFuture;
+
+
+public abstract class CollectionTypes {
+	private static final Log LOG = LoggerFactory.make( Log.class, MethodHandles.lookup() );
+
+	public static CompletionStage<Object> replace(
+			CollectionType type,
+			Object original,
+			Object target,
+			SessionImplementor session,
+			Object owner,
+			Map<Object, Object> copyCache,
+			ForeignKeyDirection foreignKeyDirection)
+			throws HibernateException {
+		return needsReplacement( type, foreignKeyDirection )
+				? replace( type, original, target, session, owner, copyCache )
+				: completedFuture( target );
+	}
+
+	public static CompletionStage<Object> replace(
+			CollectionType type,
+			Object original,
+			Object target,
+			SessionImplementor session,
+			Object owner,
+			Map<Object, Object> copyCache) throws HibernateException {
+		if ( original == null ) {
+			return replaceNullOriginal( target, session );
+		}
+		else if ( !Hibernate.isInitialized( original ) ) {
+			return replaceUninitializedOriginal( type, original, target, session, copyCache );
+		}
+		else {
+			return replaceOriginal( type, original, target, session, owner, copyCache );
+		}
+	}
+
+	// todo: make org.hibernate.type.CollectionType#replaceNullOriginal public ?
+	private static CompletionStage<Object> replaceNullOriginal(
+			Object target,
+			SessionImplementor session) {
+		if ( target == null ) {
+			return null;
```

Comment:
Method returning `CompletionStage` should never return null in principle. You can return `CompletionStages.nullFuture()`